File: final-project/Source.cpp
#include <conio.h> // need to use getch()
#include <fstream> // file handling
#include <iostream> // input/output handling
#define Line "------------------------------------------ " // macros

using namespace std;

class Carriage {
protected:
	// identifires
	string carriageType, carriageID;
	int carriagePlaces;
public:
	// setting data (name and tell)
	void setData(string carriageType, string carriageID, int carriagePlaces) {
		this->carriageType = carriageType;
		this->carriageID = carriageID;
		this->carriagePlaces = carriagePlaces;
	}

	// get name and tell
	string getCarriageType() { return carriageType; }
	string getCarriageID() { return carriageID; }
	int getCarriagePlace() { return carriagePlaces; }

	// displaying name and tell
	void Display() {
		cout << Line << endl;
		cout << "  Carriage: " << carriageID << " " << carriageType << " " << carriagePlaces << endl;
	}
};

class Passenger {
protected:
	// identifires
	string firstName, lastName, ticketType;
public:
	// setting data 
	void setData(string firstName, string lastName, string ticketType) {
		this->firstName = firstName;
		this->lastName = lastName;
		this->ticketType = ticketType;
	}

	// get name and tell
	string getFirstName() { return firstName; }
	string getFamilyName() { return lastName; }
	string getTicketType() { return ticketType; }

	// displaying name and tell
	void Display() {
		cout << Line << endl;
		cout << "  Passenger: " << firstName << " " << lastName << " " << ticketType << endl;
	}
};

// Node needed to implement linked list
class Node {
public:
	int data;
	Node* next;
};

// Linked list implementation
class LinkedList {
public:
	LinkedList() { // constructor
		head = NULL;
	}

	~LinkedList() {}; // destructor
	void addNode(int val);
	void deleteFirst();
	int getFirstElement();
	void display();

private:
	Node* head;
};

int LinkedList::getFirstElement() {
	return head->data;
}

// Add a node in linked list
void LinkedList::addNode(int val) {
	// function to add node to a list
	Node* newnode = new Node();
	newnode->data = val;
	newnode->next = NULL;
	if (head == NULL) {
		head = newnode;
	}
	else {
		Node* temp = head; // head is not NULL
		while (temp->next != NULL) {
			temp = temp->next; // go to end of list
		}
		temp->next = newnode; // linking to newnode
	}
}

// Displays the linked list
void LinkedList::display() {
	if (head == NULL) {
		cout << "Passengers list is empty!" << endl;
	}
	else {
		Node* temp = head;
		while (temp != NULL) {
			cout << temp->data << " --> ";
			temp = temp->next;
		}
		cout << endl;
	}
}

// Removing element from Linked list
void LinkedList::deleteFirst() {
	// if empty
	if (head == NULL)
		cout << "Passengers list is empty!" << endl;

	// delete the first element
	else {
		Node* temp = head;
		head = head->next;
		delete(temp);
	}
}

// Stack implementation
class Stack {
	Node* front;  // points to the head of list
public:
	Stack() {
		front = NULL;
	}
	
	void push(int); // push method to add data element
	void pop(); // pop method to remove data element
	void printStack();
};

// Inserting Data in Stack (Linked List)
void Stack::push(int d) {
	// creating a new node
	Node* temp;
	temp = new Node();
	
	temp->data = d; // setting data to it

	// add the node in front of list
	if (front == NULL)
		temp->next = NULL;
	else
		temp->next = front;
	front = temp;
}

// Removing Element from Stack (Linked List)
void Stack::pop() {
	// if empty
	if (front == NULL)
		cout << "UNDERFLOW";

	// delete the first element
	else {
		Node* temp = front;
		front = front->next;
		delete(temp);
	}
}

// Displays the linked list
void Stack::printStack() {
	if (front == NULL) {
		cout << "UNDERFLOW!";
	}
	else {
		Node* temp = front;
		while (temp != NULL) {
			cout << temp->data << " <--> ";
			temp = temp->next;
		}
		cout << endl;
	}
}

int main() {
	// identifires
	string firstName, lastName, ticketType, carriageType, carriageID; int carriagePlaces;
	int Num = 1;
	int numOfBusiness = 0, numOfFirst = 0;
	double businessSpace = 0, firstSpace = 0;

	// objects
	Passenger Psg;
	Carriage Crg;

	// start
	for (int i = 0; i < 1000; i++) { // loop for Menu
		system("cls");

		cout << "\t*** MENU *** " << endl;
		cout << "1. Add passenger to file" << endl;
		cout << "2. Read passengers info from file" << endl;
		cout << "3. Add carriage to file" << endl;
		cout << "4. Read carriage info from file" << endl;
		cout << "5. Show linked lists" << endl; // Sort and Divide Info which comes from File and will add to linked list
		cout << "Your choice:";

		switch (_getch()) {
		case 49: { // write to file a passenger
			cout << "\n" << Line << endl;
			ofstream out;
			out.open("Passengers", ios::binary | ios::app);

			cout << "\t*** Add Passenger ***" << endl;
			cout << "Enter first name: "; cin >> firstName;
			cout << "Enter last name: "; cin >> lastName;
			cout << "Enter ticket type (Business/First): "; cin >> ticketType;

			Psg.setData(firstName, lastName, ticketType);
			out.write((char*)&Psg, sizeof(Passenger));
			out.close();

			cout << endl;
			system("pause");
		}
			   break;

		case 50: { // read from file passengers
			cout << "\n\t*** Passenger File *** " << endl;

			ifstream in;
			in.open("Passengers", ios::binary);
			while (in.read((char*)&Psg, sizeof(Passenger))) {
				cout << Num << "." << endl;
				Psg.Display();
				Num++;
			}
			cout << Line << endl;
			Num = 1;
			in.close();

			cout << endl;
			system("pause");
		}
			   break;

		case 51: { // write to file carriages
			cout << "\n" << Line << endl;
			ofstream out;
			out.open("Carriages", ios::binary | ios::app);

			cout << "\t*** Add Carriages ***" << endl;
			cout << "Enter ID: "; cin >> carriageID;
			cout << "Enter type (Business/First): "; cin >> carriageType;
			cout << "Enter number of seets: "; cin >> carriagePlaces;

			Crg.setData(carriageType, carriageID, carriagePlaces);
			out.write((char*)&Crg, sizeof(Carriage));
			out.close();

			cout << endl;
			system("pause");
		}
			   break;

		case 52: { // read from file carriages
			cout << endl;
			cout << "\n\t*** Carriage File ***" << endl;
			ifstream in;
			in.open("Carriages", ios::binary);
			while (in.read((char*)&Crg, sizeof(Carriage))) {
				cout << Num << "." << endl;
				Crg.Display();
				Num++;
			}
			cout << Line << endl;
			Num = 1;
			in.close();

			cout << endl;
			system("pause");
		}
			   break;

		case 53: {
			// Sort Passengers into Linked Lists
			LinkedList* listBusiness = new LinkedList();
			LinkedList* listFirst = new LinkedList();
			numOfBusiness = 0;
			numOfFirst = 0;
			ifstream in;
			in.open("Passengers", ios::binary);
			while (in.read((char*)&Psg, sizeof(Passenger))) {
				if (Psg.getTicketType() == "Business") {
					listBusiness->addNode(Num);
					numOfBusiness++;
				}
				else {
					listFirst->addNode(Num);
					numOfFirst++;
				}
				Num++;
			}
			cout << endl << Line << endl;
			Num = 1;
			in.close();
			// ---------------------------------------
			cout << "Linked List - 'Business class'" << endl;
			listBusiness->display();
			cout << "Overall 'Business': " << numOfBusiness << endl << endl;

			cout << "Linked List - 'First class'" << endl;
			listFirst->display();
			cout << "Overall 'First': " << numOfFirst << endl;
		    // ---------------------------------------

			for (int j = 0; j < 100; j++) {
				cout << "\n\t*** INNER MENU *** " << endl;
				cout << "1. Distribute 'Business class'" << endl;
				cout << "2. Distribute 'First class'" << endl;
				cout << "0. Go back " << endl;
				cout << "Your choice:";

				businessSpace = 0;
				firstSpace = 0;

				ifstream in;
				in.open("Carriages", ios::binary);
				while (in.read((char*)&Crg, sizeof(Carriage))) {
					if (Crg.getCarriageType() == "Business") {
						businessSpace = businessSpace + Crg.getCarriagePlace();
					}
					else {
						firstSpace = firstSpace + Crg.getCarriagePlace();
					}
				}
				businessSpace = round(businessSpace / 2.0);
				firstSpace = round(firstSpace / 2.0);
				in.close();

				switch (_getch()) {
				case 49: { 
					// Distribute 'Business class'
					Stack* stackBusiness = new Stack();

					// Free spaces we have in business class
					cout << endl << Line << endl;
					cout << "Number of free spaces (Business): " << businessSpace << endl;

					if (numOfBusiness <= businessSpace) {
						// take the first element from list, PUSH() it to stack, delete that element
						for (int i = numOfBusiness; i > 0; i--) {
							int distributedPerson = listBusiness->getFirstElement();
							stackBusiness->push(distributedPerson);
							listBusiness->deleteFirst();
							businessSpace--;
						}
						// shows the number of free spaces
						cout << "Free spaces left: " << businessSpace << endl;
					}
					else if (numOfBusiness > businessSpace) {
						for (int i = businessSpace; i > 0; i--) {
							int distributedPerson = listBusiness->getFirstElement();
							stackBusiness->push(distributedPerson);
							listBusiness->deleteFirst();
						}
						cout << "You need " << numOfBusiness - businessSpace << " more spaces!" << endl;
					}
					else {
					// Overflow case
						cout << "You need " << numOfBusiness - businessSpace << " free spaces!"<< endl;
					}

					cout << "Business class passengers list: ";
					// listBusiness->display(); // displays old linked list
					stackBusiness->printStack(); // displays new stack

					cout << endl;
					system("pause");
				}
					   break;
				case 50: { // Distribute 'First class'
					Stack* stackFirst = new Stack();

					// Free spaces we have in first class
					cout << endl << Line << endl;
					cout << "Number of free spaces (Fist): " << firstSpace << endl;

					if (numOfFirst <= firstSpace) {
						// take the first element from list, PUSH() it to stack, delete that element
						for (int i = numOfFirst; i > 0; i--) {
							int distributedPerson = listFirst->getFirstElement();
							stackFirst->push(distributedPerson);
							listFirst->deleteFirst();
							firstSpace--;
						}
						// shows the number of free spaces
						cout << "Free spaces left: " << firstSpace << endl;
					}
					else if (numOfFirst > firstSpace) {
						for (int i = firstSpace; i > 0; i--) {
							int distributedPerson = listFirst->getFirstElement();
							stackFirst->push(distributedPerson);
							listFirst->deleteFirst();
						}
						cout << "You need " << numOfFirst - firstSpace << " more spaces!" << endl;
					}
					else {
						// Overflow case
						cout << "You need " << numOfFirst - firstSpace << " spaces!" << endl;
					}

					cout << "First class passengers list: ";
					stackFirst->printStack(); // displays new stack

					cout << endl;
					system("pause");
				}
					   break;
				case 48: { // go back
					j = 101;
				}
					   break;
				}
				
			} // inner menu

			cout << endl;
			// system("pause");
		}
			   break;

		default: {
			cout << "\n\nYour choice is not available in the menu!" << endl;
			cout << endl;
			system("pause");
		}
			   break;
		} // end of switch
	} // end of loop
	system("pause");
	return 0;
}

File: final-project/Source.py
import os
import math

class Carriage:
    def __init__(self):
	self.carriageType = ""
	self.carriageID = ""
	self.carriagePlaces = 0

    def set_data(self, carriageType, carriageID, carriagePlaces):
	self.carriageType = carriageType
	self.carriageID = carriageID
	self.carriagePlaces = carriagePlaces

    def get_carriage_type(self):
	return self.carriageType

    def get_carriage_id(self):
	return self.carriageID

    def get_carriage_places(self):
	return self.carriagePlaces

    def display(self):
	print("------------------------------------------ ")
	print("  Carriage: " + self.carriageID + " " + self.carriageType + " " + str(self.carriagePlaces))

class Passenger:
    def __init__(self):
	self.firstName = ""
	self.lastName = ""
	self.ticketType = ""

    def set_data(self, firstName, lastName, ticketType):
	self.firstName = firstName
	self.lastName = lastName
	self.ticketType = ticketType

    def get_first_name(self):
	return self.firstName

    def get_last_name(self):
	return self.lastName

    def get_ticket_type(self):
	return self.ticketType

    def display(self):
	print("------------------------------------------ ")
	print("  Passenger: " + self.firstName + " " + self.lastName + " " + self.ticketType)

class Node:
    def __init__(self, data):
	self.data = data
	self.next = None

class LinkedList:
    def __init__(self):
	self.head = None

    def add_node(self, val):
	new_node = Node(val)
	if self.head is None:
	    self.head = new_node
	else:
	    temp = self.head
	    while temp.next is not None:
		temp = temp.next
	    temp.next = new_node

    def delete_first(self):
	if self.head is None:
	    print("Passengers list is empty!")
	else:
	    self.head = self.head.next

    def get_first_element(self):
	if self.head is not None:
	    return self.head.data

    def display(self):
	if self.head is None:
	    print("Passengers list is empty!")
	else:
	    temp = self.head
	    while temp is not None:
		print(temp.data, end=" --> ")
		temp = temp.next
	    print()

class Stack:
    def __init__(self):
	self.front = None

    def push(self, d):
	new_node = Node(d)
	if self.front is None:
	    new_node.next = None
	else:
	    new_node.next = self.front
	    self.front = new_node

    def pop(self):
	if self.front is None:
	    print("UNDERFLOW")
	else:
	    self.front = self.front.next

    def print_stack(self):
	if self.front is None:
	    print("UNDERFLOW!")
	else:
	    temp = self.front
	    while temp is not None:
		print(temp.data, end=" <--> ")
		temp = temp.next
	    print()

def main():
    Num = 1
    numOfBusiness = 0
    numOfFirst = 0
    businessSpace = 0
    firstSpace = 0

    while True:
	os.system('cls' if os.name == 'nt' else 'clear')
	print("\t*** MENU *** ")
	print("1. Add passenger to file")
	print("2. Read passengers info from file")
	print("3. Add carriage to file")
	print("4. Read carriage info from file")
	print("5. Show linked lists")
	choice = input("Your choice:")

	if choice == "1":
	    print("\n------------------------------------------ ")
	    with open("Passengers", "ab") as out:
		print("\t*** Add Passenger ***")
		firstName = input("Enter first name: ")
		lastName = input("Enter last name: ")
		ticketType = input("Enter ticket type (Business/First): ")
		Psg = Passenger()
		Psg.set_data(firstName, lastName, ticketType)
		out.write(Psg.__dict__.encode())
	    input("\nPress Enter to continue...")

	elif choice == "2":
	    print("\n\t*** Passenger File *** ")
	    Num = 1
	    with open("Passengers", "rb") as f:
		while True:
		    data = f.read(1024)
		    if not data:
			break
		    else:
			Psg = Passenger()
			Psg.__dict__.update(eval(data))
			print(f"{Num}.")
			Psg.display()
			Num += 1
	    input("\nPress Enter to continue...")

	elif choice == "3":
	    print("\n------------------------------------------ ")
	    with open("Carriages", "ab") as out:
		print("\t*** Add Carriages ***")
		carriageID = input("Enter ID: ")
		carriageType = input("Enter type (Business/First): ")
		carriagePlaces = int(input("Enter number of seats: "))
		Crg = Carriage()
		Crg.set_data(carriageType, carriageID, carriagePlaces)
		out.write(Crg.__dict__.encode())
	    input("\nPress Enter to continue...")

	elif choice == "4":
	    print("\n\t*** Carriage File ***")
	    Num = 1
	    with open("Carriages", "rb") as f:
		while True:
		    data = f.read(1024)
		    if not data:
			break
		    else:
			Crg = Carriage()
			Crg.__dict__.update(eval(data))
			print(f"{Num}.")
			Crg.display()
			Num += 1
	    input("\nPress Enter to continue...")

	elif choice == "5":
	    listBusiness = LinkedList()
	    listFirst = LinkedList()
	    numOfBusiness = 0
	    numOfFirst = 0
	    with open("Passengers", "rb") as f:
		while True:
		    data = f.read(1024)
		    if not data:
			break
		    else:
			Psg = Passenger()
			Psg.__dict__.update(eval(data))
			if Psg.get_ticket_type() == "Business":
			    listBusiness.add_node(Num)
			    numOfBusiness += 1
			else:
			    listFirst.add_node(Num)
			    numOfFirst += 1
			Num += 1
	    print("\n------------------------------------------ ")
	    print("Linked List - 'Business class'")
	    listBusiness.display()
	    print("Overall 'Business':", numOfBusiness, "\n")
	    print("Linked List - 'First class'")
	    listFirst.display()
	    print("Overall 'First':", numOfFirst)
	    print("------------------------------------------ ")
	    businessSpace = 0
	    firstSpace = 0
	    with open("Carriages", "rb") as f:
		while True:
		    data = f.read(1024)
		    if not data:
			break
		    else:
			Crg = Carriage()
			Crg.__dict__.update(eval(data))
			if Crg.get_carriage_type() == "Business":
			    businessSpace += Crg.get_carriage_places()
			else:
			    firstSpace += Crg.get_carriage_places()
			  businessSpace = math.ceil(businessSpace / 2)
		    firstSpace = math.ceil(firstSpace / 2)
		    while True:
			print("\n\t*** INNER MENU *** ")
			print("1. Distribute 'Business class'")
			print("2. Distribute 'First class'")
			print("0. Go back ")
			inner_choice = input("Your choice:")
			if inner_choice == "1":
			    stackBusiness = Stack()
			    print("\n------------------------------------------ ")
			    print("Number of free spaces (Business):", businessSpace)
			    if numOfBusiness <= businessSpace:
				for i in range(numOfBusiness):
				    distributedPerson = listBusiness.get_first_element()
				    stackBusiness.push(distributedPerson)
				    listBusiness.delete_first()
				    businessSpace -= 1
				print("Free spaces left:", businessSpace)
			    elif numOfBusiness > businessSpace:
				for i in range(businessSpace):
				    distributedPerson = listBusiness.get_first_element()
				    stackBusiness.push(distributedPerson)
				    listBusiness.delete_first()
				print("You need", numOfBusiness - businessSpace, "more spaces!")
			    else:
				print("You need", numOfBusiness - businessSpace, "free spaces!")
			    print("Business class passengers list: ", end="")
			    stackBusiness.print_stack()
			    input("\nPress Enter to continue...")
			elif inner_choice == "2":
			    stackFirst = Stack()
			    print("\n------------------------------------------ ")
			    print("Number of free spaces (First):", firstSpace)
			    if numOfFirst <= firstSpace:
				for i in range(numOfFirst):
				    distributedPerson = listFirst.get_first_element()
				    stackFirst.push(distributedPerson)
				    listFirst.delete_first()
				    firstSpace -= 1
				print("Free spaces left:", firstSpace)
			    elif numOfFirst > firstSpace:
				for i in range(firstSpace):
				    distributedPerson = listFirst.get_first_element()
				    stackFirst.push(distributedPerson)
				    listFirst.delete_first()
				print("You need", numOfFirst - firstSpace, "more spaces!")
			    else:
				print("You need", numOfFirst - firstSpace, "spaces!")
			    print("First class passengers list: ", end="")
			    stackFirst.print_stack()
			    input("\nPress Enter to continue...")
			elif inner_choice == "0":
			    break
			else:
			    print("\nYour choice is not available in the menu!")
			    input("\nPress Enter to continue...")
	    else:
		print("\nYour choice is not available in the menu!")
		input("\nPress Enter to continue...")

if __name__ == "__main__":
    main()
